src/mono/mono/mini/mini-profiler.c

```diff
@@ -76,12 +76,12 @@ can_encode_method_ref (MonoMethod *method)
 void
 mini_profiler_emit_enter (MonoCompile *cfg)
 {
-	gboolean trace = mono_jit_trace_calls != NULL && mono_trace_eval (cfg->method);
-
-	if ((!MONO_CFG_PROFILE (cfg, ENTER) || cfg->current_method != cfg->method || (cfg->compile_aot && !can_encode_method_ref (cfg->method))) && !trace)
+	if (cfg->current_method != cfg->method)
 		return;
 
-	if (cfg->current_method != cfg->method)
+	gboolean trace = mono_jit_trace_calls != NULL && mono_trace_eval (cfg->method);
+
+	if (!trace && (!MONO_CFG_PROFILE (cfg, ENTER) || (cfg->compile_aot && !can_encode_method_ref (cfg->method))))
```

Comment:
Moving !trace to the front here could change observable behavior, did you check that MONO_CFG_PROFILE and can_encode_method_ref have no side effects?

Reply:
No side effects. This should be slightly faster.